### PR TITLE
ARTEMIS-406 STOMP acknowledgements should support transactions

### PR DESCRIPTION
**Pull request by gaohoward**
- Add support for transactional STOMP acks/nacks
- Fixed some stomp test issues

### CODE REVIEW

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompSession.java [hunk 1]
```diff
@@ -154,12 +154,16 @@ public int sendMessage(MessageReference ref,
             encoder.encode(buffer, bodySize);
             encoder.close();
          } else {
-            buffer = coreMessage.getReadOnlyBodyBuffer();
+            if (Boolean.TRUE.equals(serverMessage.getBooleanProperty(Message.HDR_LARGE_COMPRESSED))) {
+               buffer = coreMessage.getBodyBuffer();
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompSession.java]: You should always use the readOnlyBodyBuffer...

Why did you need the writeable body buffer here? it's dangerous!

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/CommandType.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      type = command;
+   }
+
+   @Override
+   public String toString() {
+      return type;
+   }
+
+   public static CommandType getType(String command) {
+      switch (command) {
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/CommandType.java]: There are serious performance implications on this, right?

you're using getType on the hot path... (on Every frame creation). What happened to that code that would inspect just a few bytes of the frame and create the StompFrame fast... this is a major regression here.

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompTransaction.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * it goes with stomp connections.
+ * it stores acks, sends, nacks
+ * during commit it applies those to core sessions.
+ * because each stomp connection uses only one session
+ * we can't rely on core session to manage multiple
+ * tx's.
+ */
+public class StompTransaction {
+
+   private List<StompAck> acks = new ArrayList<>();
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompTransaction.java]: Instead of storing these here.. Can't you do like XA is doing? Recover the transaction... and do the proper Transaction object?


Do you really need a StompTransaction? can't you reuse the actual Transaction we already have?

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompSession.java (hunk 1 above)
Comment (gaohoward) [reply artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompSession.java]: this is because we need to decompress the content before send to clients. If readonly we can't write into it.

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/CommandType.java (hunk 2 above)
Comment (gaohoward) [reply artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/CommandType.java]: ok fair point. I'll remove this.

File: artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompTransaction.java (hunk 3 above)
Comment (gaohoward) [reply artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompTransaction.java]: The StompTransaction is just a data holding acks/sends, the real transaction is delegated to core session. Using XA would be unnecessary as we are not dealing with global transactions.

Comment (clebertsuconic) [reply artemis-protocols/artemis-stomp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/stomp/StompTransaction.java]: Transactions in Artemis are usually done right away...

The operation on the journal is stored right away... then a pending commit is in place.

the thing that if you duplicate Transaction functionality from the session, you won't inherit other features such as timeouts.. etc...


If you need to support this, I would do it properly.. even if it requires some light refactoring on how transactions are stored. (I believe you could use the XID to get the transaction.. as a matter of fact the previous version had some code before such as manager.getTransactionI()).


later on when users complain about a timed out transactions, you will have a better chance of implementing through the ResourceManager that is already in place.

**Comment by clebertsuconic** [comment]
@gaohoward you mentioned you were closing this, but still open?

**Comment by gaohoward** [comment]
@clebertsuconic sorry I forgot it.
